sim/midas/src/main/cc/core/widget_registry.cc

```diff
@@ -20,5 +21,19 @@ void widget_registry_t::add_widget(bridge_driver_t *widget) {
 }
 
 void widget_registry_t::add_widget(StreamEngine *widget) {
-  stream_engine.reset(widget);
+  fprintf(stdout, "widget_registry_t::add_widget(StreamEngine)\n");
+  fprintf(stdout,
+          "cpu2fpga: %d, fpga2cpu: %d\n",
+          widget->cpu_to_fpga_cnt(),
+          widget->fpga_to_cpu_cnt());
+
+  if (stream_engine.get() == nullptr) {
+    if (fpga_stream_engine.get() != nullptr) {
+      fprintf(stdout, "PCIM stream engine driver registered before PCIS\n");
+      assert(false);
+    }
+    stream_engine.reset(widget);
+  } else if (fpga_stream_engine.get() == nullptr) {
+    fpga_stream_engine.reset(widget);
+  }
```

Comment:
Can you explain why this is needed?

Reply:
IIRC, it had to do with aligning the golden gate generated bridges and the host FPGA driver setup.
The `{CPU/FPGA}/ManagedStreamEngines` had to be configured in a particular order.